applications/FluidDynamicsApplication/python_scripts/navier_stokes_two_fluids_hydraulic_solver.py

```diff
@@ -0,0 +1,692 @@
+
+
+
+    def __CalculateTimeIntegrationAcceleration(self):
+        # Acceleration for generalised alpha time integration method.
+        alpha_m=0.5*((3-self.rho_inf)/(1+self.rho_inf))
+        alpha_f=1/(1+self.rho_inf)
+        gamma= 0.5+alpha_m-alpha_f
+        dt=self.main_model_part.ProcessInfo[KratosMultiphysics.DELTA_TIME]
+        for node in self.GetComputingModelPart().Nodes:
+            vn=node.GetSolutionStepValue(KratosMultiphysics.VELOCITY,1)
+            v = node.GetSolutionStepValue(
+                KratosMultiphysics.VELOCITY, 0)
+            acceleration_n=node.GetValue(KratosMultiphysics.ACCELERATION)
+            acceleration_alpha_method=(v-vn)/(gamma*dt)+((gamma-1)/gamma)*acceleration_n
+            node.SetValue(KratosMultiphysics.ACCELERATION,acceleration_alpha_method)
+
+    def __CalculateArtificialViscosity(self):
+
+        properties_1 = self.main_model_part.Properties[1]
+        water_dynamic_viscosity_max = self.artificial_limiter_coefficient * properties_1.GetValue(KratosMultiphysics.DYNAMIC_VISCOSITY)
+        for element in self.GetComputingModelPart().Elements:
+            artificial_viscosity_gauss = element.CalculateOnIntegrationPoints(KratosCFD.ARTIFICIAL_DYNAMIC_VISCOSITY, self.main_model_part.ProcessInfo)
+            item = 0
+            elem_artificial_viscosity= 0.0
+            for item in artificial_viscosity_gauss:
+                elem_artificial_viscosity += item
+                item +=1
+            elem_artificial_viscosity /= item
+            if elem_artificial_viscosity > water_dynamic_viscosity_max:
+                elem_artificial_viscosity = water_dynamic_viscosity_max
+            element.SetValue(KratosCFD.ARTIFICIAL_DYNAMIC_VISCOSITY, elem_artificial_viscosity)
+
+    def _PerformLevelSetConvection(self):
+        # Solve the levelset convection problem
+        self._GetLevelSetConvectionProcess().Execute()
+
+    def _PerformEulerianFmAleVelocity(self):
+
+        # Solve the historical data convection problem
+
+        domain_size = self.main_model_part.ProcessInfo[KratosMultiphysics.DOMAIN_SIZE]
+        velocity_components = [KratosMultiphysics.VELOCITY_X,KratosMultiphysics.VELOCITY_Y,KratosMultiphysics.VELOCITY_Z]
+        mesh_var = [KratosMultiphysics.MESH_VELOCITY_X,KratosMultiphysics.MESH_VELOCITY_Y,KratosMultiphysics.MESH_VELOCITY_Z]
+
+        KratosMultiphysics.VariableUtils().CopyModelPartNodalVar(KratosMultiphysics.VELOCITY,self.eulerian_convection_var, self.main_model_part, self.main_model_part, 0, 0)
+        #TODO: These 2 are not needed as the CloneSolutionStep does it
+        KratosMultiphysics.VariableUtils().CopyModelPartNodalVar(KratosMultiphysics.VELOCITY,self.eulerian_convection_var, self.main_model_part, self.main_model_part, 1, 1)
+        KratosMultiphysics.VariableUtils().CopyModelPartNodalVar(KratosMultiphysics.VELOCITY,self.eulerian_convection_var, self.main_model_part, self.main_model_part, 2, 2)
```

Comment:
Maybe I'm saying something stupid, but do we really need to copy third position of the buffer?

Reply:
No, you are right, It was done for a check during the implementation.